components/script/dom/document.rs

```diff
@@ -1575,6 +1612,97 @@ impl Document {
         ScriptThread::mark_document_with_no_blocked_loads(self);
     }
 
+    // https://html.spec.whatwg.org/multipage/#unloading-documents
+    pub fn prompt_to_unload(&self) -> bool {
+        self.incr_ignore_opens_during_unload_counter();
+        let document = Trusted::new(self);
+        let beforeunload_event = BeforeUnloadEvent::new(&self.window,
+                                           atom!("beforeunload"),
+                                           EventBubbles::Bubbles,
+                                           EventCancelable::Cancelable);
+        let event = beforeunload_event.upcast::<Event>();
+        event.set_trusted(true);
+        let event_status = self.window.upcast::<EventTarget>().dispatch_event_with_target(
+            document.root().upcast(),
+            &event,
+        );
+        // Step 7
+        if event.get_cancel_state() == EventDefault::Handled {
+            self.salvageable.set(false);
+        }
+        let mut can_unload = true;
+        if event_status == EventStatus::Canceled
+            || beforeunload_event.ReturnValue() == DOMString::from_string("".to_string()) {
+            // Step 8
+            // TODO: ask the user to confirm that they wish to unload the document.
+        }
+        self.decr_ignore_opens_during_unload_counter();
+        // Step 9
+        for iframe in self.iter_iframes() {
+            if let Some(document) = iframe.GetContentDocument() {
```

Comment:
This doesn't handle the case of cross origin iframes, we will probably need to go via the constellation here. This could end up being pretty tricky, I would be okay for this to be a followup.